kratos/mqttengine/__mqttengine.py

```diff
@@ -0,0 +1,148 @@
+        self._client.on_connect = self._connect_cb
+        for filt, cb in self._handlers.items():
+            self._client.message_callback_add(filt, cb)
+        self._keepalive_s = keepalive_s
+        self._client.connect(broker, port, keepalive_s, bind_addr)
+
+    def disconnect(self):
+        """
+        Disconnects the client
+
+        :return:
+        """
+        self._client.disconnect()
+
+    def topic_handler(self, topic_filter: str):
+        """
+        Decorator for adding topic handler
+
+        :param topic_filter: Topic filter used for matching messages
+        :return: Wrapper for the callback
+        """
+        def topic_handler_wrapper(func: callable):
+            self._handlers[topic_filter] = func
+        return topic_handler_wrapper
+
+    def exec(self):
+        """
+        Execute the engine. Never returns
+        :return:
+        """
+        self._client.loop_forever()
+
+    def start(self):
+        """
+        Start engine in the background
+
+        :return: None
+        """
+        self._client.loop_start()
+
+    def stop(self):
+        """
+        Stop engine running in the background
+
+        :return: None
+        """
+        self._client.loop_stop()
+
+    def publish(self, topic: str, payload=None, qos=0, retain=False):
+        """
+        Publish message to a topic
+
+        :param topic: Topic to publish to
+        :param payload: Message payload. None means empty payload
+        :param qos: Quality of service
+            0 - Send but don't care if message reaches the target
+            1 - Message is received at least 1 time
+            2 - Message is received exactly 1 time
+
+        :param retain: If True, will be set as "last known good" message
+            for the topic
+        :return: None
+
+        :raises NotConnectedException: Connection has not been established
+        :raises QueueFullError: Publish FIFO is full
+        """
+        res = self._client.publish(topic, payload, qos, retain)
+        if res.rc == mqtt.MQTT_ERR_SUCCESS:
+            res.wait_for_publish()
+            return
+        if res.rc == mqtt.MQTT_ERR_NO_CONN:
+            raise exceptions.NotConnectedError("Could not publish to topic %s" % topic)
+        if res.rc == mqtt.MQTT_ERR_QUEUE_SIZE:
+            raise exceptions.QueueFullError("Send buffer is full, cannot publish.")
+        raise ValueError("Unknown result code from paho-mqtt: %s" % res.rc)
+
+    def _connect_cb(self, client, userdata, flags, rc):
+        LOGGER.info("Connected with result code %s", rc)
+        for topic in self._handlers:
+            result, mid = client.subscribe(topic)
+            if result == mqtt.MQTT_ERR_SUCCESS:
+                LOGGER.info("Successfully subscribed to topic %s", topic)
+            elif result == mqtt.MQTT_ERR_NO_CONN:
+                raise exceptions.NotConnectedError("No connection established")
+            raise ValueError("Unknown return value from paho-mqtt")
```

Comment:
Should there be an else-branch here?
```suggestion
            else:
                raise ValueError("Unknown return value from paho-mqtt")
```

Reply:
https://pheanex.github.io/pylint/#R1705

Suggestion will cause a warning from pylint.

Reply:
Since we do not return after the `LOGGER.info()` call, won't that cause the control flow to always reach the `raise ValueError()` statement?

Reply:
True, It's actually a false positive from pylint, since loop continues. afterwards